packages/pigeon/lib/dart_generator.dart

```diff
@@ -521,92 +436,1264 @@ final BinaryMessenger? ${_varNamePrefix}binaryMessenger;
-          final String returnType = _makeGenericTypeArguments(func.returnType);
-          final String genericCastCall = _makeGenericCastCall(func.returnType);
-          const String accessor = '${_varNamePrefix}replyList[0]';
-          // Avoid warnings from pointlessly casting to `Object?`.
-          final String nullablyTypedAccessor =
-              returnType == 'Object' ? accessor : '($accessor as $returnType?)';
-          final String nullHandler = func.returnType.isNullable
-              ? (genericCastCall.isEmpty ? '' : '?')
-              : '!';
-          String returnStatement = 'return';
-          if (func.returnType.isEnum) {
-            if (func.returnType.isNullable) {
-              returnStatement =
-                  '$returnStatement ($accessor as int?) == null ? null : $returnType.values[$accessor! as int]';
-            } else {
-              returnStatement =
-                  '$returnStatement $returnType.values[$accessor! as int]';
-            }
-          } else if (!func.returnType.isVoid) {
-            returnStatement =
-                '$returnStatement $nullablyTypedAccessor$nullHandler$genericCastCall';
-          }
-          returnStatement = '$returnStatement;';
+      }
+    });
+  }
 
-          indent.format('''
-final List<Object?>? ${_varNamePrefix}replyList =
-\t\tawait ${_varNamePrefix}channel.send($sendArgument) as List<Object?>?;
-if (${_varNamePrefix}replyList == null) {
-\tthrow _createConnectionError(${_varNamePrefix}channelName);
-} else if (${_varNamePrefix}replyList.length > 1) {
-\tthrow PlatformException(
-\t\tcode: ${_varNamePrefix}replyList[0]! as String,
-\t\tmessage: ${_varNamePrefix}replyList[1] as String?,
-\t\tdetails: ${_varNamePrefix}replyList[2],
-\t);''');
-          // On iOS we can return nil from functions to accommodate error
-          // handling.  Returning a nil value and not returning an error is an
-          // exception.
-          if (!func.returnType.isNullable && !func.returnType.isVoid) {
-            indent.format('''
-} else if (${_varNamePrefix}replyList[0] == null) {
-\tthrow PlatformException(
-\t\tcode: 'null-error',
-\t\tmessage: 'Host platform returned null value for non-null return value.',
-\t);''');
-          }
-          indent.format('''
-} else {
-\t$returnStatement
-}''');
-        });
+  @override
+  void writeInstanceManager(
+    DartOptions generatorOptions,
+    Root root,
+    Indent indent, {
+    required String dartPackageName,
+  }) {
+    _writeProxyApiBaseClass(indent);
+
+    const String proxyApiBaseClassName = '${classNamePrefix}ProxyApiBaseClass';
+    final Iterable<String> apiHandlerSetups =
+        root.apis.whereType<AstProxyApi>().map(
+      (AstProxyApi api) {
+        return '${api.name}.${classMemberNamePrefix}setUpMessageHandlers(${classMemberNamePrefix}instanceManager: instanceManager);';
+      },
+    );
+    indent.format('''
+/// Maintains instances used to communicate with the native objects they
```

Comment:
I think we should probably put giant strings like this into a helper file, wrapped in a method that takes all the necessary variables, but without any logic ideally.